code/modules/mob/living/status_procs.dm

```diff
@@ -392,12 +392,17 @@
 		S.duration = amount + world.time
 	else if(amount > 0)
 		S = apply_status_effect(STATUS_EFFECT_SLEEPING, amount)
+		if(voluntary)
+			S.voluntary = TRUE
+	if(!voluntary && S)
+		S.voluntary = FALSE
```

Comment:
can't you just do `if(S) => S.voluntary = voluntary`

Reply:
No. Added comment as to why.
(Of course, right now we do have checks preventing us from sleeping while already sleeping, but... Better be safe.)